infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/StreamsLogEntryReader.java

```diff
@@ -84,39 +99,100 @@ boolean shouldProcess(OpaqueEntry entry) throws ReplicationReaderException {
+        }
 
-        throw new IllegalTransactionStreamsException("There are noisy streams in the transaction log entry");
+        // Skip append this entry, will process it for the next message;
+        if (currentEntrySize + currentMsgSize > maxDataSizePerMsg) {
+            return false;
+        }
+
+        return true;
     }
 
     public void setGlobalBaseSnapshot(long snapshot, long ackTimestamp) {
         globalBaseSnapshot = snapshot;
         preMsgTs = Math.max(snapshot, ackTimestamp);
         log.info("snapshot {} ackTimestamp {} preMsgTs {} seek {}", snapshot, ackTimestamp, preMsgTs, preMsgTs + 1);
-        txStream.seek(preMsgTs + 1);
+        txOpaqueStream.seek(preMsgTs + 1);
         sequence = 0;
     }
 
+
     @Override
     public LogReplicationEntry read(UUID logEntryRequestId) throws TrimmedException, IllegalTransactionStreamsException {
+        List<OpaqueEntry> opaqueEntryList = new ArrayList<>();
+        int currentEntrySize = 0;
+        int currentMsgSize = 0;
+
         try {
-            while (txStream.hasNext()) {
-                OpaqueEntry opaqueEntry = txStream.next();
-                if (!shouldProcess(opaqueEntry)) {
-                    continue;
+            while (currentMsgSize < maxDataSizePerMsg && !hasNoiseData) {
+
+                if (lastOpaqueEntry != null && shouldProcess(lastOpaqueEntry)) {
+
+                    // If the currentEntry is too big to append the current message, will skip it and
+                    // append it to the next message as the first entry.
+                    currentEntrySize = ReaderUtility.calculateOpaqueEntrySize(lastOpaqueEntry);
+
+                    if (!checkSizeOK(lastOpaqueEntry, currentMsgSize, currentEntrySize)) {
+                        break;
+                    }
+
+                    // Add the lastOpaqueEntry to the current message.
+                    opaqueEntryList.add(lastOpaqueEntry);
+                    currentMsgSize += currentEntrySize;
+                    lastOpaqueEntry = null;
+                }
+
+                if (hasNoiseData) {
+                    break;
                 }
-                LogReplicationEntry txMessage = generateMessage(opaqueEntry, logEntryRequestId);
-                return txMessage;
+
+                if (!txOpaqueStream.hasNext()) {
+                    break;
+                }
+
+                lastOpaqueEntry = txOpaqueStream.next();
             }
+
+            log.trace("Generate LogEntryDataMessage size {} with {} entries for maxDataSizePerMsg {}. lastEnry size {}",
+                    currentMsgSize, opaqueEntryList.size(), maxDataSizePerMsg, lastOpaqueEntry == null? 0 : currentEntrySize);
+
+            if (opaqueEntryList.size() == 0 && hasNoiseData) {
+                throw new IllegalTransactionStreamsException("There are noisy streams in the transaction log entry");
+            }
+
+            if (opaqueEntryList.size() == 0) {
```

Comment:
It's cleaner if generateMessageWithOpaqueEntryList accepts the empty list and returns a null. Less if conditions.